### ISSUE TITLE
API name is invalid

### ISSUE THREAD

**Issue by drewboardman** [aws-amplify/amplify-js#13199]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication, REST API

### Amplify Version

v6

### Amplify Categories

auth, api

### Backend

Other

### Environment information

<details>

```
# Put output below this line
  System:
    OS: Linux 6.8 Arch Linux
    CPU: (12) x64 AMD Ryzen 5 5600 6-Core Processor
    Memory: 6.49 GB / 15.54 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 21.7.1 - ~/.nvm/versions/node/v21.7.1/bin/node
    npm: 10.5.0 - ~/.nvm/versions/node/v21.7.1/bin/npm
  npmPackages:
    @aws-amplify/ui-react: ^6.1.6 => 6.1.6 
    @aws-amplify/ui-react-internal:  undefined ()
    @types/react: ^18.2.66 => 18.2.72 
    @types/react-dom: ^18.2.22 => 18.2.22 
    @typescript-eslint/eslint-plugin: ^7.2.0 => 7.4.0 
    @typescript-eslint/parser: ^7.2.0 => 7.4.0 
    @vitejs/plugin-react: ^4.2.1 => 4.2.1 
    aws-amplify: ^6.0.23 => 6.0.23 
    aws-amplify/adapter-core:  undefined ()
    aws-amplify/analytics:  undefined ()
    aws-amplify/analytics/kinesis:  undefined ()
    aws-amplify/analytics/kinesis-firehose:  undefined ()
    aws-amplify/analytics/personalize:  undefined ()
    aws-amplify/analytics/pinpoint:  undefined ()
    aws-amplify/api:  undefined ()
    aws-amplify/api/server:  undefined ()
    aws-amplify/auth:  undefined ()
    aws-amplify/auth/cognito:  undefined ()
    aws-amplify/auth/cognito/server:  undefined ()
    aws-amplify/auth/enable-oauth-listener:  undefined ()
    aws-amplify/auth/server:  undefined ()
    aws-amplify/data:  undefined ()
    aws-amplify/data/server:  undefined ()
    aws-amplify/datastore:  undefined ()
    aws-amplify/in-app-messaging:  undefined ()
    aws-amplify/in-app-messaging/pinpoint:  undefined ()
    aws-amplify/push-notifications:  undefined ()
    aws-amplify/push-notifications/pinpoint:  undefined ()
    aws-amplify/storage:  undefined ()
    aws-amplify/storage/s3:  undefined ()
    aws-amplify/storage/s3/server:  undefined ()
    aws-amplify/storage/server:  undefined ()
    aws-amplify/utils:  undefined ()
    eslint: ^8.57.0 => 8.57.0 
    eslint-plugin-react-hooks: ^4.6.0 => 4.6.0 
    eslint-plugin-react-refresh: ^0.4.6 => 0.4.6 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    typescript: ^5.2.2 => 5.4.3 
    vite: ^5.2.0 => 5.2.6 
  npmGlobalPackages:
    corepack: 0.25.2
    npm: 10.5.0


```

</details>

### Describe the bug

I am attempting to use the amplify-js `auth` and `api` libraries to make a cognito auth'd request to my API Gateway endpoint (which proxies to a lambda).

I am getting the error: `InvalidApiName: API name is invalid.`

I can see my console log output with the correct/expected information:

```
Making API request to apiName test with token: eyJraWQiOiJZMmtTeUQxX ....
```

I also attempted to change the name `test` to the "actual" API Gateway "Name" field: `LambdaProxyAPI`. This did not work.

### Expected behavior

I expected an api request to be made to this endpoint.

### Reproduction steps

Will be happy to provide repro steps if they are needed.

### Code Snippet

Api Configs
```typescript
export const apiGateWayName = 'test';

export const apiGateway = {
    API: {
        REST: {
            [apiGateWayName]: {
                endpoint: 'https://123abc456.execute-api.us-east-1.amazonaws.com',
                region: 'us-east-1' // Optional
            }
        }
    }
}
```
aggregated configs
```typescript
const requiredRegion = { aws_project_region: "us-east-1" } // this is required for Amplify.configure

export const config = {
    ...requiredRegion,
    ...cognitoConfig,
    ...apiGateway
};
```
`App.tsx`
```typescript
import { useEffect, useState } from 'react'
import reactLogo from './assets/react.svg'
import viteLogo from '/vite.svg'
import { Amplify } from 'aws-amplify';
import type { WithAuthenticatorProps } from '@aws-amplify/ui-react';
import { withAuthenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
import './App.css'
import { ApiCaller } from './HelloWorldRequester';
import { fetchAuthSession } from 'aws-amplify/auth'
import { config } from './amplifyconfigs/main';

Amplify.configure(config);

export function App({ signOut, user }: WithAuthenticatorProps) {
  const [count, setCount] = useState(0)

  const viteDefault = (
    <>
      <div>
        <a href="https://vitejs.dev" target="_blank">
          <img src={viteLogo} className="logo" alt="Vite logo" />
        </a>
        <a href="https://react.dev" target="_blank">
          <img src={reactLogo} className="logo react" alt="React logo" />
        </a>
      </div>
      <h1>Vite + React</h1>
      <div className="card">
        <button onClick={() => setCount((count) => count + 1)}>
          count is {count}
        </button>
        <p>
          Edit <code>src/App.tsx</code> and save to test HMR
        </p>
      </div>
      <p className="read-the-docs">
        Click on the Vite and React logos to learn more
      </p>
    </>
  )

  const [authToken, setAuthToken] = useState<string | null>(null);

  useEffect(() => {
    getIdToken().then(token => setAuthToken(token));
  }, []);

  async function getIdToken() {
    const session = await fetchAuthSession();
    return session.tokens?.idToken?.toString()!!;
  }


  const signInStuff = (
    <>
      <h1>Hello {user?.username}</h1>
      {/* conditionally render this only if you actually have a token */}
      {authToken && <ApiCaller authToken={authToken} />}
      <button onClick={signOut}>Sign out</button>
    </>
  )

  return (
    <>
      {signInStuff}
      {viteDefault}
    </>
  );
}

export default withAuthenticator(App);
```
requesting component
```typescript
import { get } from 'aws-amplify/api';
import { apiGateWayName } from './amplifyconfigs/apiGateway';

export const ApiCaller = ({ authToken }: { authToken: string }) => {
    const handleClick = async () => {
        console.log(`Making API request to apiName ${apiGateWayName} with token: ${authToken}`)
        try {
            const restOperation = get({ 
                apiName: apiGateWayName,
                path: '/helloworld',
                options: {
                    headers: {
                        Authorization: authToken
                    }
                }
            });
            const response = await restOperation.response;
            console.log('GET call succeeded: ', response);
        } catch (e) { console.log(e); }
    };
    return (
        <button onClick={handleClick}>Make API Request</button>
    );
};
```

### Log output

<details>

```
Making API request to apiName test with token: eyJraWQiOiJZMmtTeUQxXC ....
InvalidApiName: API name is invalid.
    _AmplifyError AmplifyError.ts:13
    _ApiError APIError.ts:18
    _RestApiError RestApiError.ts:6
    assertValidationError assertValidatonError.ts:11
    resolveApiUrl resolveApiUrl.ts:17
    publicHandler publicApis.ts:7
    job createCancellableOperation.ts:22
    createCancellableOperation createCancellableOperation.ts:58
    publicHandler publicApis.ts:5
    get3 publicApis.ts:30
    get4 index.ts:44
    handleClick HelloWorldRequester.tsx:8
    React 23
    <anonymous> main.tsx:
```
I also logged the full config, and it contains this section:
```json
"API": {
    "REST": {
      "test": {
        "endpoint": "https://abc123def.execute-api.us-east-1.amazonaws.com/",
        "region": "us-east-1"
      }
    }
  }
  ```

</details>

**Comment by drewboardman** [comment]
If this may help, here is a function I'm finding in the stack trace:

```javascript
export const resolveApiUrl = (amplify, apiName, path, queryParams) => {
    const urlStr = amplify.getConfig()?.API?.REST?.[apiName]?.endpoint;
    assertValidationError(!!urlStr, RestApiValidationErrorCode.InvalidApiName);
    try {
        const url = new AmplifyUrl(urlStr + path);
        if (queryParams) {
            const mergedQueryParams = new AmplifyUrlSearchParams(url.searchParams);
            Object.entries(queryParams).forEach(([key, value]) => {
                mergedQueryParams.set(key, value);
            });
            url.search = new AmplifyUrlSearchParams(mergedQueryParams).toString();
        }
        return url;
    }
    ```

**Comment by drewboardman** [comment]
I think this may have to do with the cognito configuration looking like it may be from the v5 (top level).

**Comment by cwomack** [comment]
Hello, @drewboardman 👋 and thanks for opening this issue.  It looks like your logged config has a trailing slash which might be causing issues.  The rest of your code looks proper to me, and your `path` should have the leading slash. Can you double check to see if you might be adding an extra slash into the endpoint somewhere?

**Comment by cwomack** [comment]
@drewboardman,  wanted to check in again and see if you are still experiencing this issue.  If so, do you have a screenshot of request call possibly or a log of the recovery suggestion in the error to help narrow down where this is being thrown?

**Comment by severi** [comment]
We started getting this error after trying to upgrade amplify, we were previously running 6.0.16 and apparently something has changed in the 6.0.17 release that started causing this issue. I tried to figure out the cause of the issue by reading the changelog but could not find anything that would have indicated such a change

This is also broken for us in the newest release (6.3.4). With 6.0.16 everything is working smoothly, what changed here?
